chore: rewrite peer-store submit and retrieve docs (#605)
libp2p/js-libp2p: src/peer-store/README.md

The PeerStore manages the high level operations on its inner books. Moreover, the PeerStore should be responsible for notifying interested parties of relevant events, through its Event Emitter.

## Submitting records to the PeerStore

> Several libp2p subsystems will perform operations that might gather relevant informations about peers.

Reviewer: ```suggestion
Several libp2p subsystems will perform operations that might gather relevant information about peers.
```

### Identify
- The Identify protocol automatically runs on every connection when multiplexing is enabled. The protocol will put the multiaddrs and protocols provided by the peer to the PeerStore.
- In the background, the Identify Service is also waiting for protocol change notifications of peers via the IdentifyPush protocol. Peers may leverage the `identify-push` message to communicate protocol changes to all connected peers, so that their PeerStore can be updated with the updated protocols.
- While it is currently not supported in js-libp2p, future iterations may also support the [IdentifyDelta protocol](https://github.com/libp2p/specs/pull/176).
- Taking into account that the Identify protocol records are directly from the peer, they should be considered the source of truth and weighted accordingly.

### Peer Discovery

> - Libp2p discovery protocols aim to discover new peers in the network. In a typical discovery protocol, addresses of the peer are discovered along its peer id. Once this happens, a libp2p discovery protocol should emit a `peer` event with the information of the discovered peer and this information it added to the PeerStore by libp2p.

Reviewer: ```suggestion
- Libp2p discovery protocols aim to discover new peers in the network. In a typical discovery protocol, addresses of the peer are discovered along with its peer id. Once this happens, a libp2p discovery protocol should emit a `peer` event with the information of the discovered peer and this information will be added to the PeerStore by libp2p.
```

### Dialer

> - Libp2p API supports dialing a peer through its `multiaddr`. This way, if the node is able to establish a connection with the peer listening on the given `multiaddr`, this peer and its multiaddr should be added to the PeerStore.

Reviewer: ```suggestion
- Libp2p API supports dialing a peer given a `multiaddr`, and no prior knowledge of the peer. If the node is able to establish a connection with the peer, it and its multiaddr is added to the PeerStore.
```

### DHT

> - On some DHT operations, such as finding providers for a given CID, nodes can exchange peer data as part of the query. This peer data can include unkwown peers multiaddrs and is also stored on the PeerStore.

Reviewer: ```suggestion
- On some DHT operations, such as finding providers for a given CID, nodes may exchange peer data as part of the query. This passive peer discovery should result in the DHT emitting the `peer` event in the same way [Peer Discovery](#peerdiscovery) does.
```

## Retrieving records from the PeerStore

> When the PeerStore data is updated, this information might be important for different parties. As a consequence, PeerStore is an event based component that notifies interested parties about changes. Any subsystem interested in these notifications should subscribe the PeerStore events.

Reviewer: We should also link to the events here.
```suggestion
When data in the PeerStore is updated the PeerStore will emit events based on the changes, to allow applications and other subsystems to take action on those changes. Any subsystem interested in these notifications should subscribe the PeerStore events.
```

### Peer

> - Each time a new peer is discovered, the PeerStore should emit a `peer` event, so that interested parties can leverage this peer and establish a connection with it.

Reviewer: We should link to the API readme when the events are added there

### Protocols

> - When the known protocols of a peer change, the PeerStore emits a `change:protocols` event.

Reviewer: Link to API events section when added

>   - Libp2p topologies will be particularly interested in this, so that the subsystem can open streams with relevant peers for them

Reviewer: Delete for now, it creates more questions than it provides value here.
```suggestion
```

### Multiaddrs

> - When the known listening `multiaddrs` of a peer change, the PeerStore emits a `change:multiaddrs` event.

Reviewer: Link to API events section
